src/darwin/Framework/CHIP/CHIPDeviceController.mm

```diff
@@ -93,24 +89,74 @@ - (instancetype)init
         if ([self checkForInitError:(_persistentStorageDelegateBridge != nullptr) logMsg:kErrorPersistentStorageInit]) {
             return nil;
         }
+    }
+    return self;
+}
 
-        [self getControllerNodeId];
-        errorCode = _cppCommissioner->Init(_localDeviceId, _persistentStorageDelegateBridge, _pairingDelegateBridge);
-        if ([self checkForInitError:(CHIP_NO_ERROR == errorCode) logMsg:kErrorCommissionerInit]) {
-            return nil;
-        }
+- (BOOL)isRunning
+{
+    [self.lock lock];
+    BOOL commissionerInitialzied = _cppCommissioner != nullptr;
+    [self.lock unlock];
+    return commissionerInitialzied;
+}
 
-        _chipSelectQueue = dispatch_queue_create(CHIP_SELECT_QUEUE, DISPATCH_QUEUE_SERIAL);
-        if ([self checkForInitError:(_chipSelectQueue != nil) logMsg:kErrorNetworkDispatchQueueInit]) {
-            return nil;
-        }
+- (BOOL)shutdown
+{
+    [self.lock lock];
+    if (_cppCommissioner) {
+        CHIP_LOG_DEBUG("%@", kInfoStackShutdown);
+        _cppCommissioner->Shutdown();
+        delete _cppCommissioner;
+        _cppCommissioner = nullptr;
+    }
+    [self.lock unlock];
+    return YES;
+}
 
-        // Start the IO pump
-        dispatch_async(_chipSelectQueue, ^() {
-            self.cppCommissioner->ServiceEvents();
-        });
+- (BOOL)startup:(id<CHIPPersistentStorageDelegate>)storageDelegate queue:(nonnull dispatch_queue_t)queue
+{
+    if ([self isRunning]) {
+        return YES;
```

Comment:
One way to avoid those kind of scenarios is to setup a queue for the `CHIPDeviceController`, where every API just queues up tasks serially and returns values in closures/blocks. Maybe the addition of start/stop warrants that kind of a refactor if we can't find another way to protect access

Reply:
FYI, I just refactored the controller in this PR to use a dispatch queue to make sure we don't have the controller get freed in the middle of other things. Using the queue effectively widens the critical section scope and allows returning and error handling without using labels.